server/web_push.go

```diff
@@ -0,0 +1,132 @@
+package server
+
+import (
+	"database/sql"
+
+	_ "github.com/mattn/go-sqlite3" // SQLite driver
+)
+
+// Messages cache
+const (
+	createWebPushSubscriptionsTableQuery = `
+		BEGIN;
+		CREATE TABLE IF NOT EXISTS web_push_subscriptions (
```

Comment:
1. Maybe just call the table `subscriptions`, since the entire database is about web push?
2. The updated_at column is never used
3. When do subscriptions expire, or how are they kept alive? To prevent infinite growth of this table, we have to actively prune records. I'd be okay with having a long period before pruning, but at some point, an unused record has to be deleted. Imaging thousands of people trying it out and then never using it again.

Reply:
I put the `updated_at` column in to have the possibility to add expiry logic in the future, and to do some manual testing. I'm not so sure how it should work yet.

If a user revokes permission, the push service sends a `410 Gone` which will cause our publisher to delete the subscription automatically. In practice this isn't always the case - it works immediately for Chrome, but at least Apple doesn't seem to do it.

I also added a scheduled update in the web app, so people who regularly open the web app should have a recent `updated_at` - the real problem is when people subscribe to something and then don't open ntfy. It would not be the best UX if subscriptions silently expire - so I wonder if the solution is to send a message such as "Please reopen the ntfy web app to continue receiving background notifications" after e.g. 1 month, then expire a day later.

Sort of tangentially, we could also have the service worker notify the ntfy server about receipt of a web push event and store that (e.g. some sort of `last_sent_at` and `last_received_by_web_app_at` representation). There's also a spec for getting receipts, but I don't see any implementations for it, and the service worker reporting it would be a more complete guarantee anyway https://datatracker.ietf.org/doc/html/rfc8030#section-6.3

Reply:
> I also added a scheduled update in the web app, so people who regularly open the web app should have a recent updated_at 

:+1: 

> the real problem is when people subscribe to something and then don't open ntfy.

Don't you think people would click on the notification every now and then, and that would bring them to the web app, which would then send a "I'm still alive" message? This is how I keep access tokens for the web session alive. They expire after 72 hours, but are always extended by 72 hours unless you log in. 

Tangentially (a lot of tangents...), does this access token logic^ cause issues with the PWA?

> It would not be the best UX if subscriptions silently expire - so I wonder if the solution is to send a message such as "Please reopen the ntfy web app to continue receiving background notifications" after e.g. 1 month, then expire a day later.

Assuming there are truly scenarios in which people never ever open the web app, then the UX would be bad. But don't you think people will click on the notification?

Reply:
I hate any sort of user tracking. ;-) This doesn't really serve a purpose. 

Eventually when we do a synchronized "mark as read" then it'll serve a purpose, but now if it's just for tracking, then i don't like that

Reply:
> I hate any sort of user tracking. ;-) This doesn't really serve a purpose.

Fair, though I mean tracking the service worker receiving the message - so ACK not tracking whether the user interacts or reads with it at all.